proposals/list-patterns.md

```diff
@@ -0,0 +1,142 @@
+- If the *type* is an *array_type*, the *length_pattern_clause* is disambiguated so that `int[] [0]` would match an empty integer array.
+- All other combinations are valid, for instance `T (p0, p1) [p2] { name: p3 } v` or `T (p0, p1) [p2] { p3 } v` where each clause can be omitted.
+
+> **Open question**: Should we support all these combinations?
+
+#### Pattern compatibility
+
+A *length_pattern* is compatible with any type that is *countable*, i.e. has an accessible property getter that returns an `int` and has the name `Length` or `Count`. If both properties are present, the former is preferred.
+
+A *list_pattern* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the *length_pattern*
+2. Has an accessible indexer with a single `int` parameter
+
+ > **Open question**: Should we support `this[Index]` indexers? If so, which one is preferred if `this[int]` is also present?
+
+A *slice_pattern* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the *length_pattern*
+2. Has an accessible `Slice` method that takes two `int` parameters (required only if a subpattern is specified)
+
+ > **Open question**: Should we support `this[Range]` indexers? If so, which one is preferred if `Slice(int, int)` is also present?
+
+This set of rules is already specified as the [***range indexer pattern***](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md#implicit-index-support) and required for range indexers.
+
+#### Subsumption checking
+
+Subsumption checking works just like positional patterns with `ITuple` - corresponding subpatterns are matched by position plus an additional node for testing length.
+
+#### Lowering
+
+A pattern of the form `expr is {1, 2, 3}` is equivalent to the following code:
+```cs
+expr.Length is 3
+&& expr[0] is 1
+&& expr[1] is 2
+&& expr[2] is 3
+```
+A *slice_pattern* acts like a proper discard i.e. no tests will be emitted for such pattern, rather it only affects other nodes, namely the length and indexer. For instance, a pattern of the form `expr is {1, .. var s, 3}`  is equivalent to the following code:
+```cs
+expr.Length    is >= 2
+&& expr[0]     is 1
+&& expr[1..^1] is var s
+&& expr[^1]    is 3
+```
+The *input type* for the *slice_pattern* is the return type of the underlying `Slice` method with two exceptions: For `string` and arrays, `string.Substring` and `RuntimeHelpers.GetSubArray` will be used, respectively.
+
+Note: the lowering is presented in the pattern form here to show how subsumption checking works, for example, the following code produces an error because both patterns yield the same DAG:
+
+```cs
+case {_, .., 1}: // expr.Length is >= 2 && expr[^1] is 1
+case {.., _, 1}: // expr.Length is >= 2 && expr[^1] is 1
+```
+Unlike:
+```cs
+case {_, 1, ..}: // expr.Length is >= 2 && expr[1] is 1
+case {.., 1, _}: // expr.Length is >= 2 && expr[^2] is 1
```

Comment:
Just curious, would we detect the redundancy if we know the length? (my assumption is _no_)
```cs
case [3] {_, 1, ..}: // expr.Length is >= 2 && expr[1] is 1
case [3] {.., 1, _}: // expr.Length is >= 2 && expr[^2] is 1
```

Reply:
The correct codegen for that example checks `Length=3` but we wouldn't warn or anything about those redundancies.

(The effective length range is defined as the intersection of the length pattern value set and the inferred length value set from the list, if this results in an empty value set we produce an error.)